File: samples/vdsp_op/cvtcolor/cvtcolor.py
import vaststreamx as vsx

import cv2
import argparse
import numpy as np
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../../..")
sys.path.append(common_path)
import common.utils as utils

attr = vsx.AttrKey


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--input_file",
        default="data/images/dog.jpg",
        help="input file",
    )
    parser.add_argument(
        "--output_file",
        default="cvtcolor_result.jpg",
        help="output file",
    )
    parser.add_argument(
        "--cvtcolor_code",
        default="bgr2rgb_interleave2planar",
        help="cvtcolor code",
    )
    args = parser.parse_args()
    return args


def parse_cvtcolor_code(cvtcolor_code):
    if cvtcolor_code.upper() == "YUV2RGB_NV12":
        return (
            vsx.ImageFormat.YUV_NV12,
            vsx.ImageFormat.RGB_PLANAR,
            vsx.ColorCvtCode.YUV2RGB_NV12,
        )
    elif cvtcolor_code.upper() == "YUV2BGR_NV12":
        return (
            vsx.ImageFormat.YUV_NV12,
            vsx.ImageFormat.BGR_PLANAR,
            vsx.ColorCvtCode.YUV2BGR_NV12,
        )
    elif cvtcolor_code.upper() == "BGR2RGB":
        return (
            vsx.ImageFormat.BGR_PLANAR,
            vsx.ImageFormat.RGB_PLANAR,
            vsx.ColorCvtCode.BGR2RGB,
        )
    elif cvtcolor_code.upper() == "RGB2BGR":
        return (
            vsx.ImageFormat.RGB_PLANAR,
            vsx.ImageFormat.BGR_PLANAR,
            vsx.ColorCvtCode.RGB2BGR,
        )
    elif cvtcolor_code.upper() == "BGR2RGB_INTERLEAVE2PLANAR":
        return (
            vsx.ImageFormat.BGR_INTERLEAVE,
            vsx.ImageFormat.RGB_PLANAR,
            vsx.ColorCvtCode.BGR2RGB_INTERLEAVE2PLANAR,
        )
    elif cvtcolor_code.upper() == "RGB2BGR_INTERLEAVE2PLANAR":
        return (
            vsx.ImageFormat.RGB_INTERLEAVE,
            vsx.ImageFormat.BGR_PLANAR,
            vsx.ColorCvtCode.RGB2BGR_INTERLEAVE2PLANAR,
        )
    elif cvtcolor_code.upper() == "BGR2BGR_INTERLEAVE2PLANAR":
        return (
            vsx.ImageFormat.BGR_INTERLEAVE,
            vsx.ImageFormat.BGR_PLANAR,
            vsx.ColorCvtCode.BGR2BGR_INTERLEAVE2PLANAR,
        )
    elif cvtcolor_code.upper() == "RGB2RGB_INTERLEAVE2PLANAR":
        return (
            vsx.ImageFormat.RGB_INTERLEAVE,
            vsx.ImageFormat.RGB_PLANAR,
            vsx.ColorCvtCode.RGB2RGB_INTERLEAVE2PLANAR,
        )
    elif cvtcolor_code.upper() == "YUV2GRAY_NV12":
        return (
            vsx.ImageFormat.YUV_NV12,
            vsx.ImageFormat.GRAY,
            vsx.ColorCvtCode.YUV2GRAY_NV12,
        )
    elif cvtcolor_code.upper() == "BGR2GRAY_INTERLEAVE":
        return (
            vsx.ImageFormat.BGR_INTERLEAVE,
            vsx.ImageFormat.GRAY,
            vsx.ColorCvtCode.BGR2GRAY_INTERLEAVE,
        )
    elif cvtcolor_code.upper() == "BGR2GRAY_PLANAR":
        return (
            vsx.ImageFormat.BGR_PLANAR,
            vsx.ImageFormat.GRAY,
            vsx.ColorCvtCode.BGR2GRAY_PLANAR,
        )
    elif cvtcolor_code.upper() == "RGB2GRAY_INTERLEAVE":
        return (
            vsx.ImageFormat.RGB_INTERLEAVE,
            vsx.ImageFormat.GRAY,
            vsx.ColorCvtCode.RGB2GRAY_INTERLEAVE,
        )
    elif cvtcolor_code.upper() == "RGB2GRAY_PLANAR":
        return (
            vsx.ImageFormat.RGB_PLANAR,
            vsx.ImageFormat.GRAY,
            vsx.ColorCvtCode.RGB2GRAY_PLANAR,
        )
    elif cvtcolor_code.upper() == "RGB2YUV_NV12_PLANAR":
        return (
            vsx.ImageFormat.RGB_PLANAR,
            vsx.ImageFormat.YUV_NV12,
            vsx.ColorCvtCode.RGB2YUV_NV12_PLANAR,
        )
    elif cvtcolor_code.upper() == "BGR2YUV_NV12_PLANAR":
        return (
            vsx.ImageFormat.BGR_PLANAR,
            vsx.ImageFormat.YUV_NV12,
            vsx.ColorCvtCode.BGR2YUV_NV12_PLANAR,
        )
    else:
        assert False, f"Unsuport cvtcolor code: {cvtcolor_code}"


if __name__ == "__main__":
    args = argument_parser()

    input_format, output_format, cvtcolor_code = parse_cvtcolor_code(args.cvtcolor_code)

    vsx.set_device(args.device_id)

    cv_image = cv2.imread(args.input_file)
    assert cv_image is not None, f"Failed to read: {args.input_file}"

    vsx_image = utils.cv_bgr888_to_vsximage(cv_image, input_format, args.device_id)

    # Build operator
    op = vsx.BuildInOperator(vsx.OpType.SINGLE_OP_CVT_COLOR)

    op.set_attribute(
        {
            attr.IIMAGE_WIDTH: vsx_image.width,
            attr.IIMAGE_HEIGHT: vsx_image.height,
            attr.IIMAGE_WIDTH_PITCH: vsx_image.width,
            attr.IIMAGE_HEIGHT_PITCH: vsx_image.height,
            attr.OIMAGE_WIDTH: vsx_image.width,
            attr.OIMAGE_HEIGHT: vsx_image.height,
            attr.OIMAGE_WIDTH_PITCH: vsx_image.width,
            attr.OIMAGE_HEIGHT_PITCH: vsx_image.height,
            attr.COLOR_CVT_CODE: cvtcolor_code,
            attr.COLOR_SPACE: vsx.ImageColorSpace.COLOR_SPACE_BT601,
        }
    )

    # run
    outputs = op.execute(
        [vsx_image], [(output_format, vsx_image.width, vsx_image.height)]
    )

    output = outputs[0]

    out_image_cv = utils.vsximage_to_cv_bgr888(outputs[0])

    # write
    cv2.imwrite(args.output_file, out_image_cv)
    print(f"Write result to args.output_file")

File: samples/vdsp_op/cvtcolor/cvtcolor.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include "common/cmdline.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"
#include "vaststreamx/vaststreamx.h"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("input_file", '\0', "input image", false,
                        "../data/images/dog.jpg");
  args.add<std::string>("output_file", '\0', "output image", false,
                        "cvtcolor_result.jpg");
  args.add<std::string>("cvtcolor_code", '\0', "cvtcolor code", false,
                        "bgr2rgb_interleave2planar");
  args.parse_check(argc, argv);
  return args;
}

vsx::BuildInOperatorAttrColorCvtCode GetCvtColorCode(
    const std::string& cvtcolor_code, vsx::ImageFormat& input_format,
    vsx::ImageFormat& output_format) {
  std::string code_upper = cvtcolor_code;
  std::transform(code_upper.begin(), code_upper.end(), code_upper.begin(),
                 [](unsigned char c) { return std::toupper(c); });

  if (code_upper == "YUV2RGB_NV12") {
    input_format = vsx::YUV_NV12;
    output_format = vsx::RGB_PLANAR;
    return vsx::kYUV2RGB_NV12;
  }
  if (code_upper == "YUV2BGR_NV12") {
    input_format = vsx::YUV_NV12;
    output_format = vsx::BGR_PLANAR;
    return vsx::kYUV2BGR_NV12;
  }

  if (code_upper == "BGR2RGB") {
    input_format = vsx::BGR_PLANAR;
    output_format = vsx::RGB_PLANAR;
    return vsx::kBGR2RGB;
  }
  if (code_upper == "RGB2BGR") {
    input_format = vsx::RGB_PLANAR;
    output_format = vsx::BGR_PLANAR;
    return vsx::kRGB2BGR;
  }
  if (code_upper == "BGR2RGB_INTERLEAVE2PLANAR") {
    input_format = vsx::BGR_INTERLEAVE;
    output_format = vsx::RGB_PLANAR;
    return vsx::kBGR2RGB_INTERLEAVE2PLANAR;
  }
  if (code_upper == "RGB2BGR_INTERLEAVE2PLANAR") {
    input_format = vsx::RGB_INTERLEAVE;
    output_format = vsx::BGR_PLANAR;
    return vsx::kRGB2BGR_INTERLEAVE2PLANAR;
  }
  if (code_upper == "BGR2BGR_INTERLEAVE2PLANAR") {
    input_format = vsx::BGR_INTERLEAVE;
    output_format = vsx::BGR_PLANAR;
    return vsx::kBGR2BGR_INTERLEAVE2PLANAR;
  }
  if (code_upper == "RGB2RGB_INTERLEAVE2PLANAR") {
    input_format = vsx::RGB_INTERLEAVE;
    output_format = vsx::RGB_PLANAR;
    return vsx::kRGB2RGB_INTERLEAVE2PLANAR;
  }
  if (code_upper == "YUV2GRAY_NV12") {
    input_format = vsx::YUV_NV12;
    output_format = vsx::GRAY;
    return vsx::kYUV2GRAY_NV12;
  }
  if (code_upper == "BGR2GRAY_INTERLEAVE") {
    input_format = vsx::BGR_INTERLEAVE;
    output_format = vsx::GRAY;
    return vsx::kBGR2GRAY_INTERLEAVE;
  }
  if (code_upper == "BGR2GRAY_PLANAR") {
    input_format = vsx::BGR_PLANAR;
    output_format = vsx::GRAY;
    return vsx::kBGR2GRAY_PLANAR;
  }
  if (code_upper == "RGB2GRAY_INTERLEAVE") {
    input_format = vsx::RGB_INTERLEAVE;
    output_format = vsx::GRAY;
    return vsx::kRGB2GRAY_INTERLEAVE;
  }
  if (code_upper == "RGB2GRAY_PLANAR") {
    input_format = vsx::RGB_PLANAR;
    output_format = vsx::GRAY;
    return vsx::kRGB2GRAY_PLANAR;
  }
  if (code_upper == "RGB2YUV_NV12_PLANAR") {
    input_format = vsx::RGB_PLANAR;
    output_format = vsx::YUV_NV12;
    return vsx::kRGB2YUV_NV12_PLANAR;
  }
  if (code_upper == "BGR2YUV_NV12_PLANAR") {
    input_format = vsx::BGR_PLANAR;
    output_format = vsx::YUV_NV12;
    return vsx::kBGR2YUV_NV12_PLANAR;
  }

  CHECK(false) << "Unrecognize cvtcolor code: " << cvtcolor_code << std::endl;
}

int main(int argc, char* argv[]) {
  auto args = ArgumentParser(argc, argv);
  auto device_id = args.get<uint32_t>("device_id");

  vsx::ImageFormat input_format, output_format;
  auto cvtcolor_code = GetCvtColorCode(args.get<std::string>("cvtcolor_code"),
                                       input_format, output_format);
  vsx::SetDevice(device_id);

  vsx::Image image;
  vsx::MakeVsxImage(args.get<std::string>("input_file"), image, input_format);
  auto op =
      vsx::BuildInOperator(vsx::BuildInOperatorType::kSINGLE_OP_CVT_COLOR);

  op.SetAttribute<vsx::AttrKey::kIimageWidth>(image.Width());
  op.SetAttribute<vsx::AttrKey::kIimageHeight>(image.Height());
  op.SetAttribute<vsx::AttrKey::kIimageWidthPitch>(image.Width());
  op.SetAttribute<vsx::AttrKey::kIimageHeightPitch>(image.Height());
  op.SetAttribute<vsx::AttrKey::kOimageWidth>(image.Width());
  op.SetAttribute<vsx::AttrKey::kOimageHeight>(image.Height());
  op.SetAttribute<vsx::AttrKey::kOimageWidthPitch>(image.Width());
  op.SetAttribute<vsx::AttrKey::kOimageHeightPitch>(image.Height());
  op.SetAttribute<vsx::AttrKey::kColorCvtCode>(cvtcolor_code);
  op.SetAttribute<vsx::AttrKey::kColorSpace>(vsx::kCOLOR_SPACE_BT601);

  // copy input to device memory
  auto input_vacc = vsx::Image(image.Format(), image.Width(), image.Height(),
                               vsx::Context::VACC(device_id));
  input_vacc.CopyFrom(image);

  // make device memory image for output
  auto output_vacc = vsx::Image(output_format, image.Width(), image.Height(),
                                vsx::Context::VACC(device_id));
  std::vector<vsx::Image> outputs{output_vacc};

  // buildin operator execute
  op.Execute({input_vacc}, outputs);

  // copy result to cpu memory
  vsx::Image output_cpu = vsx::Image(output_vacc.Format(), output_vacc.Width(),
                                     output_vacc.Height());
  output_cpu.CopyFrom(output_vacc);

  cv::Mat out_mat;
  vsx::ConvertVsxImageToCvMatBgrPacked(output_cpu, out_mat);
  cv::imwrite(args.get<std::string>("output_file"), out_mat);
  std::cout << "Save result to " << args.get<std::string>("output_file")
            << std::endl;
  return 0;
}